crates/icrate/examples/metal.rs

```diff
@@ -0,0 +1,344 @@
+    rc::Id,
+    runtime::ProtocolObject,
+    ClassType,
+};
+
+#[repr(C)]
+pub struct SceneProperties {
+    pub time: f32,
+}
+
+#[repr(C)]
+pub struct VertexInput {
+    pub position: Position,
+    pub color: Color,
+}
+
+#[repr(C)]
+pub struct Position {
+    pub x: f32,
+    pub y: f32,
+    pub z: f32,
+}
+
+#[repr(C)]
+pub struct Color {
+    pub r: f32,
+    pub g: f32,
+    pub b: f32,
+}
+
+// declare the Objective-C class machinery
+declare_class!(
+    // declare the delegate class with our instance variables
+    struct Delegate {
+        device: IvarDrop<Id<ProtocolObject<dyn MTLDevice>>, "_device">,
+        command_queue: IvarDrop<Id<ProtocolObject<dyn MTLCommandQueue>>, "_command_queue">,
+        pipeline_state: IvarDrop<Id<ProtocolObject<dyn MTLRenderPipelineState>>, "_pipeline_state">,
+        pipeline_descriptor: IvarDrop<Id<MTLRenderPipelineDescriptor>, "_pipeline_descriptor">,
+        window: IvarDrop<Id<NSWindow>, "_window">,
+        mtk_view: IvarDrop<Id<MTKView>, "_mtk_view">,
+        start_date: IvarDrop<Id<NSDate>, "_start_date">,
+    }
+    mod ivars;
+
+    // declare the class type
+    unsafe impl ClassType for Delegate {
+        type Super = NSObject;
+        type Mutability = InteriorMutable;
+        const NAME: &'static str = "Delegate";
+    }
+
+    // define the delegate methods for the `NSApplicationDelegate` protocol
+    unsafe impl NSApplicationDelegate for Delegate {
+        #[method(applicationDidFinishLaunching:)]
+        #[allow(non_snake_case)]
+        unsafe fn applicationDidFinishLaunching(&self, _notification: &NSNotification) {
+            // configure the metal view delegate
+            unsafe {
+                let object = ProtocolObject::from_ref(self);
+                self.mtk_view.setDelegate(Some(object));
+            }
+
+            // configure the window
+            unsafe {
+                self.window.setContentView(Some(&self.mtk_view));
+                self.window.center();
+                self.window.setTitle(ns_string!("metal example"));
+                self.window.makeKeyAndOrderFront(None);
+            }
+        }
+    }
+
+    // define the Delegate methods (e.g., initializer)
+    unsafe impl Delegate {
+        #[method(initWithShaders:)]
+        #[allow(non_snake_case)]
+        unsafe fn __initWithShaders(this: *mut Self, shaders: &NSString) -> Option<NonNull<Self>> {
+            let this: Option<&mut Self> = msg_send![super(this), init];
+
+            // get the default device
+            let device = {
+                let ptr = unsafe { MTLCreateSystemDefaultDevice() };
+                unsafe { Id::retain(ptr) }.expect("Failed to get default system device.")
+            };
+
+            // create the command queue
+            let command_queue = device
+                .newCommandQueue()
+                .expect("Failed to create a command queue.");
+
+            // create the app window
+            let window = {
```

Comment:
This is sadly usually the wrong way to do it; windows should be initialized in or after `(applicationDidFinishLaunching:`, and unsure about this, but the same probably goes for a lot of the other metal stuff.

At the very least, a comment about this would be nice.

Reply:
Yeah, I wasn't sure about this either.

One thing that might be a problem though is how do I actually modify the ivars in that case since `applicationDidFinishLaunching` takes `&self`?

I saw the comments for the recent changes about interior mutability and using `Cell` or `RefCell` but when I tried an example earlier I couldn't figure out how to get anything with that to actually work.

Can you maybe give an example for how I might be able to do that? (What would need to change with the struct definition, etc.)

Reply:
I think you'd have to do `window: IvarDrop<Box<RefCell<Option<Id<NSWindow>>>>, "_window">`, however ugly that is :/

I'm working on getting rid of the `Box`, but the `RefCell` and the `Option` are basically unavoidable due to how Rust ownership works. I might consider making a helper tool for this very common case of "initialize later", but maybe I'll also just defer it to a higher-level crate like [`cacao`](https://github.com/ryanmcgrath/cacao).